src/System.IO.Pipelines/ReadableBufferReader.cs

```diff
@@ -80,7 +86,7 @@ public int Take()
 
         private void MoveNext()
         {
-            if (_enumerator.MoveNext())
+            if (_enumerator.MoveNext() && !_enumerator.Current.IsEmpty)
```

Comment:
Should probably be a while like inside of the ctor. If you have an empty segment and then you peek. It will explode.